File: DONGUK/프로그래머스_WEEK_15/코딩 테스트 공부.py
import math

INF = math.inf

def solution(alp, cop, problems):
    answer = 0
    
    max_alp, max_cop = -1, -1
    for problem in problems:
        cmp_alp, cmp_cop = problem[0], problem[1]
        
        if cmp_alp > max_alp:
            max_alp = cmp_alp
        
        if cmp_cop > max_cop:
            max_cop = cmp_cop
            
    
    if alp > max_alp:
        max_alp = alp
    
    if cop > max_cop:
        max_cop = cop
            
    
    dp = [[INF] * (max_cop + 1) for _ in range(max_alp + 1)]
    
    dp[alp][cop] = 0
    for i in range(alp, max_alp + 1):
        for j in range(cop, max_cop + 1):
            
            if i + 1 <= max_alp:
                dp[i+1][j] = min(dp[i+1][j], dp[i][j] + 1)
                
            if j + 1 <= max_cop:
                dp[i][j+1] = min(dp[i][j+1], dp[i][j] + 1)
            
    
            for problem in problems:
                req_alp, req_cop = problem[0], problem[1]
                alp_rwd, cop_rwd = problem[2], problem[3]
                cost = problem[4]
                
                if i >= req_alp and j >= req_cop:
                    next_alp = min(max_alp, i + alp_rwd)
                    next_cop = min(max_cop, j + cop_rwd)
                    
                    dp[next_alp][next_cop] = min(dp[next_alp][next_cop], dp[i][j] + cost)
        
        
    answer = dp[-1][-1]
                
    return answer

File: JINSEOP/프로그래머스_WEEK15/코딩 테스트 공부.cpp
#include <string>
#include <vector>
using namespace std;
#define INF 1000000000

int DP[151][151];
int max_alp = 0, max_cop = 0;
int study(int alp, int cop, vector<vector<int>>& problems) {
    if (alp >= max_alp && cop >= max_cop) return 0;

    if (alp > max_alp) alp = max_alp;
    if (cop > max_cop) cop = max_cop;

    int& res = DP[alp][cop];
    if (res != 0) return res;
    res = INF;

    for (auto vec : problems) {
        if (alp < vec[0] || cop < vec[1]) continue;
        res = min(res, study(alp + vec[2], cop + vec[3], problems) + vec[4]);
    }

    res = min(res, study(alp + 1, cop, problems) + 1);
    res = min(res, study(alp, cop + 1, problems) + 1);

    return res;
}

int solution(int alp, int cop, vector<vector<int>> problems) {
    int answer = 0;
    
    for (auto vec : problems) {
        max_alp = max(max_alp, vec[0]);
        max_cop = max(max_cop, vec[1]);
    }
    
    answer = study(alp, cop, problems);
    return answer;
}

// 재귀와 DP에 대해서 좀 더 생각해볼 것!
